### PR TITLE
Added DiagnosticSource for filters

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#3279 

@rynowak @avanderhoorn

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs [hunk 1]
```diff
@@ -241,6 +241,17 @@ private IFilterMetadata[] GetFilters()
...
+            if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Mvc.RegisteredFilters"))
+            {
+                _diagnosticSource.Write(
+                    "Microsoft.AspNet.Mvc.RegisteredFilters",
+                    new
+                    {
+                        actionDescriptor = ActionContext.ActionDescriptor,
+                        filterProviderContext = context
+                    });
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: This is totally out of place, what is this for?

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs [hunk 2]
```diff
@@ -301,8 +312,30 @@ private Task InvokeAllAuthorizationFiltersAsync()
...
             else if (current.Filter != null)
             {
+                if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Mvc.BeforeOnAuthorization"))
+                {
+                    _diagnosticSource.Write(
+                        "Microsoft.AspNet.Mvc.BeforeOnAuthorization",
+                        new
+                        {
+                            actionDescriptor = _authorizationContext.ActionDescriptor,
+                            authorizationContext = _authorizationContext
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: Include the filter here as well

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs [hunk 3]
```diff
@@ -357,10 +390,34 @@ private Task InvokeAllResourceFiltersAsync()
...
 
+                    if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Mvc.AfterOnResourceExecution"))
+                    {
+                        _diagnosticSource.Write(
+                            "Microsoft.AspNet.Mvc.AfterOnResourceExecution",
+                            new
+                            {
+                                actionDescriptor = _resourceExecutingContext.ActionDescriptor,
+                                resourceExecutingContext = _resourceExecutingContext,
+                                resourceExecutedContext = _resourceExecutedContext
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: Isn't this always null here?

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: IN all of these

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs [hunk 4]
```diff
@@ -357,10 +390,34 @@ private Task InvokeAllResourceFiltersAsync()
...
                         InvokeResourceFilterAsync);
 
+                    if (_diagnosticSource.IsEnabled("Microsoft.AspNet.Mvc.AfterOnResourceExecution"))
+                    {
+                        _diagnosticSource.Write(
+                            "Microsoft.AspNet.Mvc.AfterOnResourceExecution",
+                            new
+                            {
+                                actionDescriptor = _resourceExecutingContext.ActionDescriptor,
+                                resourceExecutingContext = _resourceExecutingContext,
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: I think you can leave this out in the 'after'

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs (hunk 1 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: At this point the `FilterProviderContext` contains a list of `FIlterItem` which provides the FilterDescriptors and Filters

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: but what is this for?

Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: I thought @avanderhoorn needed the list of registered filters separately.

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs (hunk 4 above)
Comment (ajaybhargavb) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: Won't it have the updated `Result` property?

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: This event on it's own is out of place, all of these other events already provide this.

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: `resourceExecutedContext` also has the updated `.Result` property

**Comment by ajaybhargavb** [comment]
Updated.

**Comment by avanderhoorn** [comment]
In terms of working out `ExceptionFilter` vs `ActionFilter` vs `ResultFilter` etc is the recommended approach here to derive this from the `type` of the filter?

**Comment by rynowak** [comment]
> In terms of working out ExceptionFilter vs ActionFilter vs ResultFilter etc is the recommended approach here to derive this from the type of the filter?

From the name of the event - if you want to consolidate that and track it as 'generic filter' feel free.

**Comment by avanderhoorn** [comment]
Thanks

**Comment by ajaybhargavb** [comment]
Updated (with extension methods)

File: src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs [hunk 5]
```diff
@@ -288,8 +289,18 @@ private Task InvokeAllAuthorizationFiltersAsync()
             var current = _cursor.GetNextFilter<IAuthorizationFilter, IAsyncAuthorizationFilter>();
             if (current.FilterAsync != null)
             {
+                _diagnosticSource.BeforeOnAuthorizationAsync(
+                    _authorizationContext.ActionDescriptor,
+                    _authorizationContext,
+                    current.FilterAsync);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Controllers/FilterActionInvoker.cs]: Simplify this method to just take the `_authorizationContext` and `current.FilterAsync`

**Comment by rynowak** [comment]
:shipit: after removing a parameter from all of the methods :+1:

**Pull request merged**